Java05. ДЗ 03 Кузиванов Николай

Carups opened the pull request.

Carups committed:
complete task 3

Carups commented:
@dsavvinov

dsavvinov commented on src/main/java/ru/spbau/mit/TrieImpl.java:
```diff
@@ -0,0 +1,143 @@
+    private Node root = new Node();
+
+    @Override
+    public String toString() {
```
Рекомендуется делать `toString()` таким, чтобы он занимал какое-то гарантированно малое время и был более-менее side-effect free. 
Рекомендация происходит от того, что `toString()` активно используется, например, в дебаггерах для рендеринга элементов, и нетривиальные `toString()` могут замедлить работу/изменить состояние программы на пустом месте. 

В данном случае, можно, например, ввести отдельный метод `String render()`

dsavvinov commented on src/main/java/ru/spbau/mit/TrieImpl.java:
```diff
@@ -0,0 +1,143 @@
+    @Override
+    public String toString() {
+        StringBuilder answerString = new StringBuilder();
+        LinkedList<Node> queue = new LinkedList<>();
```
Minor: В Java есть интерфейс `Queue` с методами типа `poll` (и `LinkedList` его удачно имплементит)

dsavvinov commented on src/main/java/ru/spbau/mit/TrieImpl.java:
```diff
@@ -0,0 +1,143 @@
+        StringBuilder answerString = new StringBuilder();
+        LinkedList<Node> queue = new LinkedList<>();
+        queue.add(root);
+        while (queue.size() != 0)
```
Minor: сравнения `size()` с нулем лучше заменить на `isEmpty()`

dsavvinov commented on src/main/java/ru/spbau/mit/TrieImpl.java:
```diff
@@ -0,0 +1,143 @@
+        LinkedList<Node> queue = new LinkedList<>();
+        queue.add(root);
+        while (queue.size() != 0)
+        {
```
Code-style: скобки все еще "прыгают"

dsavvinov commented on src/main/java/ru/spbau/mit/TrieImpl.java:
```diff
@@ -0,0 +1,143 @@
+            answerString.append(currentNode.toString());
+            Object keys[] = currentNode.next.keySet().toArray();
+            Arrays.sort(keys);
+            for (Object c :
```
Code-style: не надо разбивать на несколько строк конструкцию, если она помещается в одну

dsavvinov commented on src/main/java/ru/spbau/mit/Node.java:
```diff
@@ -0,0 +1,27 @@
+    @Override
+    public String toString() {
+        StringBuilder resultString = new StringBuilder();
+        resultString.append(";");
```
Minor: `append` возвращает `this` (т.е. тот же самый `StringBuilder`), поэтому обычно вместо нескольких вызовов подряд пишут цепочку вызовов append:
`resultString.append(";").append(Boolean.toString(isTerminal) + ":").append(...)`

dsavvinov replied:
Все эти символы-разделители -- это те же самые магические константы (только теперь строковые, а не числовые). Их следует выносить в отдельные `static final`-переменные с внятными названиями. В частности, тогда код в десериализации начинает выглядеть более понятно, и не будет вызывать вопросов типа: "Ой, а что это за двоеточие тут с неба упало?"

dsavvinov commented on src/main/java/ru/spbau/mit/TrieImpl.java:
```diff
@@ -0,0 +1,143 @@
+            Node currentNode = queue.getFirst();
+            queue.removeFirst();
+            answerString.append(currentNode.toString());
+            Object keys[] = currentNode.next.keySet().toArray();
```
Зачем нарочно загрублять тип? Тут же статически известно, что это `Character[]`.

dsavvinov replied:
Code-style, здесь и далее: несмотря на то, что Java позволяет писать скобки массива после имени переменной (как в C), так никто никогда не делает, и все пишут после типа: `Object[] keys`. 

Мотивация такая, что `[]` -- это часть типа, а не имени переменной

dsavvinov commented on src/main/java/ru/spbau/mit/TrieImpl.java:
```diff
@@ -0,0 +1,143 @@
+
+        for(int i = sourceForNodes.length - 1; i > 0; --i)
+        {
+
```
Code-style: лишняя пустая строка (судя по консистентности появления таких пустых строк, это намеренное решение, но оно не очень удачное -- такие пустые строки никак не помогают читаемости кода, но с успехом "едят" вертикальное место)

dsavvinov commented on src/main/java/ru/spbau/mit/TrieImpl.java:
```diff
@@ -0,0 +1,143 @@
+        {
+
+            String sourceNode[] = sourceForNodes[i].split(":");
+            Node curNode = new Node();
```
Когда работаешь с серализацией/десериализацией, очень хорошо бы постараться достигнуть максимальной симметричности между логикой, собственно, сериализации и десериализации. Это значительно упрощает чтение и поддержку такого кода: сразу видно, что вот, мы тут ожидаем увидеть boolean потому что вон там его пишем.

Это к тому, что вот тут кусок явно про десериализацию `Node`, и хорошо бы, что бы он жил в `Node` рядом с `toString`.

dsavvinov commented on src/main/java/ru/spbau/mit/TrieImpl.java:
```diff
@@ -0,0 +1,143 @@
+                c = dis.readChar();
+                inputStringBuilder.append(c);
+            }
+        } catch (EOFException e){
```
Как уже обсуждалось, исключения должны быть исключительной ситуацией в программе, а не частью логики.

А вообще в `DataInput/OutputStream` есть просто `read/writeUTF`, который делает ровно то, что нужно, только нормально)

dsavvinov commented on src/main/java/ru/spbau/mit/TrieImpl.java:
```diff
@@ -0,0 +1,143 @@
+            if (sourceNode.length > 2) {
+                for (int key = sourceNode[2].length() - 1; key >= 0; --key) {
+                    curNode.next.put(sourceNode[2].charAt(key), queue.getFirst());
+                    queue.removeFirst();
```
Я бы, конечно, по возможности, избегал всех этих stateful-штук. Причина проста: про них сложно думать. Вот например, глядя на этот код, сходу абсолютно неочевидно, что в `queue` всегда будет какая-нибудь вершина, и уж тем более, что там будет именно та вершина, которая соответствует этому символу.

Способов избежать этого в конкретной ситуации два: или изменить формат сериализации/десериализации, чтобы в нем все было более явно (например, если писать дерево в порядке обхода DFSа, то алгоритм десериализации получается stateless), или если уж очень надо, то писать много документации: что тут происходит, какие инварианты поддерживаются на каждой итерации, и т.д.

dsavvinov reviewed (changes requested):
1. У сериализации в строки есть еще существенный недостаток: сериализованное представление получается очень жирное. Это не то чтобы критично, просто хотелось бы, чтобы было понимание этого.

2. Если уж изобретаешь свой формат сериализации, то придется задуматься о разных нетривиальных штуках. 
Одной из таких вещей является escaping.
В данном решении он просто забыт, и поэтому можно сложить в дерево строчки, содержащие `:` или `;`, после чего дерево сериализуется в кривой формат, и при попытке десериализовать такое дерево все эпично взорвется.

3. Над идиоматичностью Java-кода еще надо поработать. Я понимаю, что после плюсов тяжело переучиться, но это абсолютно doable, ничего тут невыполнимого нет. Например, можно начать с того, чтобы перестали болтаться скобки, появляться пробелы перед списком параметров функции, пустые строчки, неожиданные переносы строки, и т.д.

Надо поправить хотя бы п.2, т.к. это бага. Исправление остальных замечаний, как обычно, по желанию, но благотворно влияет на итоговую оценку.

dsavvinov commented:
@Carups Нужно исправить замечания, дедлайн: 27.05, 23:59

Carups committed:
fixes

Carups commented:
Исправил почти все указанные баги. 

"Зачем нарочно загрублять тип? Тут же статически известно, что это Character[]." Кажется что не известно, потому что не компилируется иначе. 
 

>Вторник, 24 апреля 2018, 21:17 +03:00 от Dmitry Savvinov <notifications@github.com>:
>
>@dsavvinov requested changes on this pull request.
>* 
>У сериализации в строки есть еще существенный недостаток: сериализованное представление получается очень жирное. Это не то чтобы критично, просто хотелось бы, чтобы было понимание этого.
>* 
>Если уж изобретаешь свой формат сериализации, то придется задуматься о разных нетривиальных штуках.
>Одной из таких вещей является escaping.
>В данном решении он просто забыт, и поэтому можно сложить в дерево строчки, содержащие : или ;, после чего дерево сериализуется в кривой формат, и при попытке десериализовать такое дерево все эпично взорвется.
>* 
>Над идиоматичностью Java-кода еще надо поработать. Я понимаю, что после плюсов тяжело переучиться, но это абсолютно doable, ничего тут невыполнимого нет. Например, можно начать с того, чтобы перестали болтаться скобки, появляться пробелы перед списком параметров функции, пустые строчки, неожиданные переносы строки, и т.д.
>Надо поправить хотя бы п.2, т.к. это бага. Исправление остальных замечаний, как обычно, по желанию, но благотворно влияет на итоговую оценку.
>----------------------------------------------------------------------
>In  src/main/java/ru/spbau/mit/TrieImpl.java :
>> @@ -0,0 +1,143 @@
>+package ru.spbau.mit;
>+
>+import java.io.*;
>+import java.util.Arrays;
>+import java.util.LinkedList;
>+
>+public class TrieImpl implements Trie, StreamSerializable {
>+    private Node root = new Node();
>+
>+    @Override
>+    public String toString() {
>
>Рекомендуется делать toString() таким, чтобы он занимал какое-то гарантированно малое время и был более-менее side-effect free.
>Рекомендация происходит от того, что toString() активно используется, например, в дебаггерах для рендеринга элементов, и нетривиальные toString() могут замедлить работу/изменить состояние программы на пустом месте.
>В данном случае, можно, например, ввести отдельный метод String render()
>----------------------------------------------------------------------
>In  src/main/java/ru/spbau/mit/TrieImpl.java :
>> @@ -0,0 +1,143 @@
>+package ru.spbau.mit;
>+
>+import java.io.*;
>+import java.util.Arrays;
>+import java.util.LinkedList;
>+
>+public class TrieImpl implements Trie, StreamSerializable {
>+    private Node root = new Node();
>+
>+    @Override
>+    public String toString() {
>+        StringBuilder answerString = new StringBuilder();
>+        LinkedList<Node> queue = new LinkedList<>();
>
>Minor: В Java есть интерфейс Queue с методами типа poll (и LinkedList его удачно имплементит)
>----------------------------------------------------------------------
>In  src/main/java/ru/spbau/mit/TrieImpl.java :
>> @@ -0,0 +1,143 @@
>+package ru.spbau.mit;
>+
>+import java.io.*;
>+import java.util.Arrays;
>+import java.util.LinkedList;
>+
>+public class TrieImpl implements Trie, StreamSerializable {
>+    private Node root = new Node();
>+
>+    @Override
>+    public String toString() {
>+        StringBuilder answerString = new StringBuilder();
>+        LinkedList<Node> queue = new LinkedList<>();
>+        queue.add(root);
>+        while (queue.size() != 0)
>
>Minor: сравнения size() с нулем лучше заменить на isEmpty()
>----------------------------------------------------------------------
>In  src/main/java/ru/spbau/mit/TrieImpl.java :
>> +package ru.spbau.mit;
>+
>+import java.io.*;
>+import java.util.Arrays;
>+import java.util.LinkedList;
>+
>+public class TrieImpl implements Trie, StreamSerializable {
>+    private Node root = new Node();
>+
>+    @Override
>+    public String toString() {
>+        StringBuilder answerString = new StringBuilder();
>+        LinkedList<Node> queue = new LinkedList<>();
>+        queue.add(root);
>+        while (queue.size() != 0)
>+        {
>
>Code-style: скобки все еще "прыгают"
>----------------------------------------------------------------------
>In  src/main/java/ru/spbau/mit/TrieImpl.java :
>> +public class TrieImpl implements Trie, StreamSerializable {
>+    private Node root = new Node();
>+
>+    @Override
>+    public String toString() {
>+        StringBuilder answerString = new StringBuilder();
>+        LinkedList<Node> queue = new LinkedList<>();
>+        queue.add(root);
>+        while (queue.size() != 0)
>+        {
>+            Node currentNode = queue.getFirst();
>+            queue.removeFirst();
>+            answerString.append(currentNode.toString());
>+            Object keys[] = currentNode.next.keySet().toArray();
>+            Arrays.sort(keys);
>+            for (Object c :
>
>Code-style: не надо разбивать на несколько строк конструкцию, если она помещается в одну
>----------------------------------------------------------------------
>In  src/main/java/ru/spbau/mit/Node.java :
>> +package ru.spbau.mit;
>+
>+import com.sun.corba.se.spi.ior.ObjectKey;
>+
>+import java.lang.reflect.Array;
>+import java.util.HashMap;
>+import java.util.Arrays;
>+
>+public class Node {
>+    HashMap<Character, Node> next = new HashMap<>();
>+    int availableTerminals = 0;
>+    boolean isTerminal = false;
>+    @Override
>+    public String toString() {
>+        StringBuilder resultString = new StringBuilder();
>+        resultString.append(";");
>
>Minor: append возвращает this (т.е. тот же самый StringBuilder), поэтому обычно вместо нескольких вызовов подряд пишут цепочку вызовов append:
>resultString.append(";").append(Boolean.toString(isTerminal) + ":").append(...)
>----------------------------------------------------------------------
>In  src/main/java/ru/spbau/mit/TrieImpl.java :
>> +import java.util.LinkedList;
>+
>+public class TrieImpl implements Trie, StreamSerializable {
>+    private Node root = new Node();
>+
>+    @Override
>+    public String toString() {
>+        StringBuilder answerString = new StringBuilder();
>+        LinkedList<Node> queue = new LinkedList<>();
>+        queue.add(root);
>+        while (queue.size() != 0)
>+        {
>+            Node currentNode = queue.getFirst();
>+            queue.removeFirst();
>+            answerString.append(currentNode.toString());
>+            Object keys[] = currentNode.next.keySet().toArray();
>
>Зачем нарочно загрублять тип? Тут же статически известно, что это Character[].
>----------------------------------------------------------------------
>In  src/main/java/ru/spbau/mit/TrieImpl.java :
>> +import java.util.LinkedList;
>+
>+public class TrieImpl implements Trie, StreamSerializable {
>+    private Node root = new Node();
>+
>+    @Override
>+    public String toString() {
>+        StringBuilder answerString = new StringBuilder();
>+        LinkedList<Node> queue = new LinkedList<>();
>+        queue.add(root);
>+        while (queue.size() != 0)
>+        {
>+            Node currentNode = queue.getFirst();
>+            queue.removeFirst();
>+            answerString.append(currentNode.toString());
>+            Object keys[] = currentNode.next.keySet().toArray();
>
>Code-style, здесь и далее: несмотря на то, что Java позволяет писать скобки массива после имени переменной (как в C), так никто никогда не делает, и все пишут после типа: Object[] keys.
>Мотивация такая, что [] -- это часть типа, а не имени переменной
>----------------------------------------------------------------------
>In  src/main/java/ru/spbau/mit/TrieImpl.java :
>> +        }
>+        return answerString.toString();
>+    }
>+
>+    public void serialize(OutputStream out) throws IOException {
>+        DataOutputStream dos = new DataOutputStream(out);
>+        dos.writeChars(this.toString());
>+    }
>+
>+    private void fromString(String source){
>+        LinkedList<Node> queue = new LinkedList<>();
>+        String sourceForNodes[] = source.split(";");
>+
>+        for(int i = sourceForNodes.length - 1; i > 0; --i)
>+        {
>+
>
>Code-style: лишняя пустая строка (судя по консистентности появления таких пустых строк, это намеренное решение, но оно не очень удачное -- такие пустые строки никак не помогают читаемости кода, но с успехом "едят" вертикальное место)
>----------------------------------------------------------------------
>In  src/main/java/ru/spbau/mit/Node.java :
>> +package ru.spbau.mit;
>+
>+import com.sun.corba.se.spi.ior.ObjectKey;
>+
>+import java.lang.reflect.Array;
>+import java.util.HashMap;
>+import java.util.Arrays;
>+
>+public class Node {
>+    HashMap<Character, Node> next = new HashMap<>();
>+    int availableTerminals = 0;
>+    boolean isTerminal = false;
>+    @Override
>+    public String toString() {
>+        StringBuilder resultString = new StringBuilder();
>+        resultString.append(";");
>
>Все эти символы-разделители -- это те же самые магические константы (только теперь строковые, а не числовые). Их следует выносить в отдельные static final-переменные с внятными названиями. В частности, тогда код в десериализации начинает выглядеть более понятно, и не будет вызывать вопросов типа: "Ой, а что это за двоеточие тут с неба упало?"
>----------------------------------------------------------------------
>In  src/main/java/ru/spbau/mit/TrieImpl.java :
>> +    }
>+
>+    public void serialize(OutputStream out) throws IOException {
>+        DataOutputStream dos = new DataOutputStream(out);
>+        dos.writeChars(this.toString());
>+    }
>+
>+    private void fromString(String source){
>+        LinkedList<Node> queue = new LinkedList<>();
>+        String sourceForNodes[] = source.split(";");
>+
>+        for(int i = sourceForNodes.length - 1; i > 0; --i)
>+        {
>+
>+            String sourceNode[] = sourceForNodes[i].split(":");
>+            Node curNode = new Node();
>
>Когда работаешь с серализацией/десериализацией, очень хорошо бы постараться достигнуть максимальной симметричности между логикой, собственно, сериализации и десериализации. Это значительно упрощает чтение и поддержку такого кода: сразу видно, что вот, мы тут ожидаем увидеть boolean потому что вон там его пишем.
>Это к тому, что вот тут кусок явно про десериализацию Node, и хорошо бы, что бы он жил в Node рядом с toString.
>----------------------------------------------------------------------
>In  src/main/java/ru/spbau/mit/TrieImpl.java :
>> +
>+        }
>+        root = queue.getFirst();
>+    }
>+
>+    public void deserialize(InputStream in) throws IOException {
>+        DataInputStream dis = new DataInputStream(in);
>+        char c;
>+        StringBuilder inputStringBuilder = new StringBuilder();
>+        try {
>+            while (true) {
>+
>+                c = dis.readChar();
>+                inputStringBuilder.append(c);
>+            }
>+        } catch (EOFException e){
>
>Как уже обсуждалось, исключения должны быть исключительной ситуацией в программе, а не частью логики.
>А вообще в DataInput/OutputStream есть просто read/writeUTF, который делает ровно то, что нужно, только нормально)
>----------------------------------------------------------------------
>In  src/main/java/ru/spbau/mit/TrieImpl.java :
>> +
>+    private void fromString(String source){
>+        LinkedList<Node> queue = new LinkedList<>();
>+        String sourceForNodes[] = source.split(";");
>+
>+        for(int i = sourceForNodes.length - 1; i > 0; --i)
>+        {
>+
>+            String sourceNode[] = sourceForNodes[i].split(":");
>+            Node curNode = new Node();
>+            curNode.isTerminal = Boolean.parseBoolean(sourceNode[0]);
>+            curNode.availableTerminals = Integer.parseInt(sourceNode[1]);
>+            if (sourceNode.length > 2) {
>+                for (int key = sourceNode[2].length() - 1; key >= 0; --key) {
>+                    curNode.next.put(sourceNode[2].charAt(key), queue.getFirst());
>+                    queue.removeFirst();
>
>Я бы, конечно, по возможности, избегал всех этих stateful-штук. Причина проста: про них сложно думать. Вот например, глядя на этот код, сходу абсолютно неочевидно, что в queue всегда будет какая-нибудь вершина, и уж тем более, что там будет именно та вершина, которая соответствует этому символу.
>Способов избежать этого в конкретной ситуации два: или изменить формат сериализации/десериализации, чтобы в нем все было более явно (например, если писать дерево в порядке обхода DFSа, то алгоритм десериализации получается stateless), или если уж очень надо, то писать много документации: что тут происходит, какие инварианты поддерживаются на каждой итерации, и т.д.
>—
>You are receiving this because you authored the thread.
>Reply to this email directly,  view it on GitHub , or  mute the thread .


-- 
Nicolay Kuzivanov

dsavvinov commented:
> Кажется что не известно, потому что не компилируется иначе. 

Ну потому что `toArray()` возвращает `Object[]`, т.к. иначе ему пришлось бы создавать массив `T[]`, что невозможно. Поэтому там есть вторая перегрузка, `toArray(T[])`, которая заполняет переданный массив

dsavvinov commented:
@Carups извини, я что-то не понимаю, как исправлена проблема с экранированием символов. Вроде как в бор все еще можно сложить "foo\nbar", оно успешно сериализуется в строчку, а потом десериализуется не в то, что было.

Carups committed:
Update Node.java

Carups committed:
Update TrieImpl.java

Carups commented:
@dsavvinov  так сойдёт?

dsavvinov commented:
Халтура, конечно. Ладно, зачитываю, 7.5б
